### PR TITLE
[FL-3040] Audio support for SubGhz

### PR DESCRIPTION
**Pull request by Skorpionm**
# What's new

- [FL-3040] Furi_hal_speaker: fix multiple resource usage
- SubGhz: add furi_hal_subghz_set_debug_pin
- SubGhz: add sound SubGhz Read, SubGhz Read RAW

# Verification 

- Run SubGhz Read or Read RAW, go to the settings and set Sound <ON> in the menu, go to the program's work screen, the sound of the signal should appear

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

[FL-3040]: https://flipperzero.atlassian.net/browse/FL-3040?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: firmware/targets/f7/api_symbols.csv [hunk 1]
```diff
@@ -1,5 +1,5 @@
 entry,status,name,type,params
-Version,+,10.1,,
+Version,+,13.1,,
```
Comment (DrZlo13) [inline firmware/targets/f7/api_symbols.csv]: No need to increase the major version, the minor version is enough.

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 2]
```diff
@@ -1,23 +1,57 @@
...
+#include <furi_hal_cortex.h>
+
+#define TAG "FuriHalSpeaker"
 
 #define FURI_HAL_SPEAKER_TIMER TIM16
 #define FURI_HAL_SPEAKER_CHANNEL LL_TIM_CHANNEL_CH1
 #define FURI_HAL_SPEAKER_PRESCALER 500
 #define FURI_HAL_SPEAKER_MAX_VOLUME 60
 
+FuriMutex* furi_hal_speaker_bus_mutex = NULL;
```
Comment (DrZlo13) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c]: furi_hal_speaker_mutex

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 3]
```diff
@@ -1,23 +1,57 @@
...
+
 // #define FURI_HAL_SPEAKER_NEW_VOLUME
 
 void furi_hal_speaker_init() {
+    furi_hal_speaker_bus_mutex = furi_mutex_alloc(FuriMutexTypeNormal);
```
Comment (DrZlo13) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c]: assert that furi_hal_speaker_bus_mutex is NULL, or free if it's not null

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 4]
```diff
@@ -1,23 +1,57 @@
...
     FURI_CRITICAL_ENTER();
     LL_TIM_DeInit(FURI_HAL_SPEAKER_TIMER);
     FURI_CRITICAL_EXIT();
+    FURI_LOG_I(TAG, "Init OK");
+}
+
+void furi_hal_speaker_deinit() {
+    LL_TIM_DeInit(FURI_HAL_SPEAKER_TIMER);
+    furi_hal_gpio_init(&gpio_speaker, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+    furi_mutex_free(furi_hal_speaker_bus_mutex);
```
Comment (DrZlo13) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c]: furi_hal_speaker_bus_mutex = NULL;

File: firmware/targets/furi_hal_include/furi_hal_speaker.h [hunk 5]
```diff
@@ -4,12 +4,22 @@
...
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 void furi_hal_speaker_init();
 
+void furi_hal_speaker_deinit();
+
+bool __attribute__((warn_unused_result)) furi_hal_speaker_acquire(uint32_t timeout);
```
Comment (DrZlo13) [inline firmware/targets/furi_hal_include/furi_hal_speaker.h]: Ownership workflow needs to be documented. It is important to clarify that ownership is transferred to the thread, and also to explain how ownership works in interrups.

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 6]
```diff
@@ -54,6 +88,8 @@ static inline uint32_t furi_hal_speaker_calculate_compare(float volume) {
 }
 
 void furi_hal_speaker_start(float frequency, float volume) {
+    furi_check(furi_mutex_get_owner(furi_hal_speaker_bus_mutex) == furi_thread_get_current_id());
```
Comment (DrZlo13) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c]: What about interrupts?

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 7]
```diff
@@ -75,6 +111,7 @@ void furi_hal_speaker_start(float frequency, float volume) {
 }
 
 void furi_hal_speaker_set_volume(float volume) {
+    furi_check(furi_mutex_get_owner(furi_hal_speaker_bus_mutex) == furi_thread_get_current_id());
```
Comment (DrZlo13) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c]: Interrupts?

**Review by DrZlo13 — CHANGES_REQUESTED** [review]

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c (hunk 2 above)
Comment (DrZlo13) [reply firmware/targets/f7/furi_hal/furi_hal_speaker.c]: also, static

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c (hunk 3 above)
Comment (skotopes) [reply firmware/targets/f7/furi_hal/furi_hal_speaker.c]: No need

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c (hunk 6 above)
Comment (skotopes) [reply firmware/targets/f7/furi_hal/furi_hal_speaker.c]: also you already have furi_hal_speaker_is_mine which can be reused here.

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 8]
```diff
@@ -1,23 +1,57 @@
...
+void furi_hal_speaker_release() {
+    furi_check(furi_mutex_get_owner(furi_hal_speaker_bus_mutex) == furi_thread_get_current_id());
+    furi_hal_speaker_stop();
+    furi_hal_gpio_init(&gpio_speaker, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+    furi_hal_power_insomnia_exit();
+    furi_check(furi_mutex_release(furi_hal_speaker_bus_mutex) == FuriStatusOk);
+}
+
+bool furi_hal_speaker_is_mine() {
+    return furi_mutex_get_owner(furi_hal_speaker_bus_mutex) == furi_thread_get_current_id();
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c]: If FURI_IS_IRQ_MODE - return true.

File: firmware/targets/f7/furi_hal/furi_hal_subghz.c [hunk 9]
```diff
@@ -17,39 +17,26 @@
...
- *      gpio_ext_pa4
- *      gpio_ext_pa6
- *      gpio_ext_pa7
- * Attention this setting switches pin to output. 
- * Make sure it is not connected directly to power or ground
- */
-
-//#define SUBGHZ_DEBUG_CC1101_PIN gpio_ext_pa7
-#ifdef SUBGHZ_DEBUG_CC1101_PIN
 uint32_t subghz_debug_gpio_buff[2];
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_subghz.c]: static and proper name please

File: firmware/targets/f7/furi_hal/furi_hal_subghz.c [hunk 10]
```diff
@@ -372,6 +359,40 @@ void furi_hal_subghz_set_path(FuriHalSubGhzPath path) {
...
+        }
+    } else if(furi_hal_subghz.gpio_debug_pin != NULL) {
+        furi_hal_gpio_init(
+            furi_hal_subghz.gpio_debug_pin, GpioModeOutputPushPull, GpioPullNo, GpioSpeedVeryHigh);
+        ret = true;
+    }
+    return ret;
+}
+
+static bool furi_hal_subghz_stop_degug() {
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_subghz.c]: degug -> debug

File: firmware/targets/f7/furi_hal/furi_hal_subghz.c [hunk 11]
```diff
@@ -372,6 +359,40 @@ void furi_hal_subghz_set_path(FuriHalSubGhzPath path) {
...
+    bool ret = false;
+    if(furi_hal_subghz.gpio_debug_pin == &gpio_speaker) {
+        if(furi_hal_speaker_is_mine()) {
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_subghz.c]: furi_hal_speaker is none of subghz business

File: firmware/targets/f7/furi_hal/furi_hal_subghz.c [hunk 12]
```diff
@@ -673,30 +691,27 @@ bool furi_hal_subghz_start_async_tx(FuriHalSubGhzAsyncTxCallback callback, void*
...
-    dma_config.PeriphRequest = LL_DMAMUX_REQ_TIM2_UP;
-    dma_config.Priority = LL_DMA_PRIORITY_VERYHIGH;
-    LL_DMA_Init(DMA1, LL_DMA_CHANNEL_2, &dma_config);
-    LL_DMA_SetDataLength(DMA1, LL_DMA_CHANNEL_2, 2);
-    LL_DMA_EnableChannel(DMA1, LL_DMA_CHANNEL_2);
-
-#endif
+    // Start debug
+    if(furi_hal_subghz_start_debug()) {
+        const GpioPin* gpio = furi_hal_subghz.gpio_debug_pin;
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_subghz.c]: why this section is not in furi_hal_subghz_start_debug

File: firmware/targets/f7/furi_hal/furi_hal_subghz.c [hunk 13]
```diff
@@ -730,10 +745,10 @@ void furi_hal_subghz_stop_async_tx() {
     // Deinitialize GPIO
     furi_hal_gpio_init(&gpio_cc1101_g0, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
 
-#ifdef SUBGHZ_DEBUG_CC1101_PIN
-    LL_DMA_DisableChannel(DMA1, LL_DMA_CHANNEL_2);
-    furi_hal_gpio_init(&SUBGHZ_DEBUG_CC1101_PIN, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
-#endif
+    // Stop debug
+    if(furi_hal_subghz_stop_degug()) {
+        LL_DMA_DisableChannel(DMA1, LL_DMA_CHANNEL_2);
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_subghz.c:738]: same thing

File: firmware/targets/furi_hal_include/furi_hal_speaker.h (hunk 5 above)
Comment (skotopes) [reply firmware/targets/furi_hal_include/furi_hal_speaker.h]: Yep, docs please for all methods in furi_hal_speaker

File: firmware/targets/furi_hal_include/furi_hal_subghz.h [hunk 14]
```diff
@@ -60,6 +61,28 @@ typedef enum {
     SubGhzRegulationTxRx, /**TxRx*/
 } SubGhzRegulation;
 
+/*
+ * IO GO0 CC1101 to an external comb.
+ * Debug pin can be assigned
```
Comment (skotopes) [inline firmware/targets/furi_hal_include/furi_hal_subghz.h]: any pin actually, not only externals

File: applications/plugins/music_player/music_player_worker.c [hunk 15]
```diff
@@ -47,47 +47,52 @@ static int32_t music_player_worker_thread_callback(void* context) {
...
-                    instance->callback_context);
-            }
-
-            furi_hal_speaker_stop();
-            furi_hal_speaker_start(frequency, volume);
-            while(instance->should_work && furi_get_tick() < next_tick) {
-                volume *= 0.9945679;
-                furi_hal_speaker_set_volume(volume);
-                furi_delay_ms(2);
+    if(furi_hal_speaker_acquire(1000)) {
```
Comment (skotopes) [inline applications/plugins/music_player/music_player_worker.c:50]: 30ms is enough

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 16]
```diff
@@ -1,23 +1,59 @@
...
+        furi_hal_gpio_init_ex(
+            &gpio_speaker, GpioModeAltFunctionPushPull, GpioPullNo, GpioSpeedLow, GpioAltFn14TIM16);
+        return true;
+    } else {
+        return false;
+    }
+}
+
+void furi_hal_speaker_release() {
+    furi_check(furi_mutex_get_owner(furi_hal_speaker_mutex) == furi_thread_get_current_id());
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c]: furi_hal_speaker_is_mine

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 17]
```diff
@@ -1,23 +1,59 @@
...
+    furi_hal_speaker_stop();
+    furi_hal_gpio_init(&gpio_speaker, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+    furi_hal_power_insomnia_exit();
+    furi_check(furi_mutex_release(furi_hal_speaker_mutex) == FuriStatusOk);
+}
+
+bool furi_hal_speaker_is_mine() {
+    return furi_mutex_get_owner(furi_hal_speaker_mutex) == furi_thread_get_current_id();
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c]: if FURI_IS_IRQ_MODE return true

**Review by skotopes — CHANGES_REQUESTED** [review]

File: firmware/targets/furi_hal_include/furi_hal_speaker.h (hunk 5 above)
Comment (DrZlo13) [reply firmware/targets/furi_hal_include/furi_hal_speaker.h]: process != thread, what about the corner case, call in interrupts?

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 18]
```diff
@@ -1,23 +1,60 @@
...
     LL_TIM_DeInit(FURI_HAL_SPEAKER_TIMER);
     FURI_CRITICAL_EXIT();
+    FURI_LOG_I(TAG, "Init OK");
+}
+
+void furi_hal_speaker_deinit() {
+    furi_check(furi_hal_speaker_mutex = NULL);
+    LL_TIM_DeInit(FURI_HAL_SPEAKER_TIMER);
+    furi_hal_gpio_init(&gpio_speaker, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+    furi_mutex_free(furi_hal_speaker_mutex);
```
Comment (DrZlo13) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c:33]: furi_hal_speaker_mutex = NULL; after free

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 19]
```diff
@@ -1,23 +1,60 @@
...
+    furi_assert(furi_hal_speaker_mutex == NULL);
+    furi_hal_speaker_mutex = furi_mutex_alloc(FuriMutexTypeNormal);
     FURI_CRITICAL_ENTER();
     LL_TIM_DeInit(FURI_HAL_SPEAKER_TIMER);
     FURI_CRITICAL_EXIT();
+    FURI_LOG_I(TAG, "Init OK");
+}
+
+void furi_hal_speaker_deinit() {
+    furi_check(furi_hal_speaker_mutex = NULL);
```
Comment (DrZlo13) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c]: wtf?

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 20]
```diff
@@ -1,23 +1,60 @@
...
+    LL_TIM_DeInit(FURI_HAL_SPEAKER_TIMER);
+    furi_hal_gpio_init(&gpio_speaker, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+    furi_mutex_free(furi_hal_speaker_mutex);
+}
 
-    furi_hal_gpio_init_ex(
-        &gpio_speaker, GpioModeAltFunctionPushPull, GpioPullNo, GpioSpeedLow, GpioAltFn14TIM16);
+bool furi_hal_speaker_acquire(uint32_t timeout) {
```
Comment (DrZlo13) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c:37]: always return true if in isr before function code

File: firmware/targets/f7/furi_hal/furi_hal_speaker.c [hunk 21]
```diff
@@ -1,23 +1,60 @@
...
+        furi_hal_power_insomnia_enter();
+        furi_hal_gpio_init_ex(
+            &gpio_speaker, GpioModeAltFunctionPushPull, GpioPullNo, GpioSpeedLow, GpioAltFn14TIM16);
+        return true;
+    } else {
+        return false;
+    }
+}
+
+void furi_hal_speaker_release() {
```
Comment (DrZlo13) [inline firmware/targets/f7/furi_hal/furi_hal_speaker.c:50]: return if in isr before function code

**Review by DrZlo13 — DISMISSED** [review]

**Review by skotopes — APPROVED** [review]

File: firmware/targets/f7/api_symbols.csv (hunk 1 above)
Comment (skotopes) [reply firmware/targets/f7/api_symbols.csv]: Actually there is a need. speaker api now behaves differently.

**Pull request merged**